src/main/scala/firrtl/passes/wiring/WiringUtils.scala

```diff
@@ -84,86 +97,132 @@ object WiringUtils {
+        .map{ v =>
+          owners(v) = owners(v) ++ owners(u)
+          queue.enqueue(v)
+        }
+    }
 
-  /** Sets the sharedParent of lineage top
-    */
-  def setSharedParent(top: String)(lin: Lineage): Lineage = lin map setSharedParent(top) match {
-    case l if l.name == top => l.copy(sharedParent = true)
-    case l => l
+    val sinkInsts = i.fullHierarchy.keys
+      .filter { case WDefInstance(_, _, module, _) =>
+        sinks.map(namedName(_)).contains(module) }
+      .map    { k => i.fullHierarchy(k)                                      }
+      .toSeq.flatten
+
+    // Check that every sink has one unique owner. The only time that
+    // this should fail is if a sink is equidistant to two sources.
+    sinkInsts.map { s =>
+      if (!owners.contains(s) || owners(s).size > 1) {
+        throw new WiringException(s"Unable to determine source mapping for sink '${s.map(_.name)}'") }
+    }
+
+    owners
+      .filter { case (k, v) => sinkInsts.contains(k) }
+      .map    { case (k, v) => (k, v.flatten)        }
+      .toMap
   }
 
-  /** Sets the addPort and cons fields of the lineage tree
-    */
-  def setThings(portNames:Map[String, String], compName: String)(lin: Lineage): Lineage = {
-    val funs = Seq(
-      ((l: Lineage) => l map setThings(portNames, compName)),
-      ((l: Lineage) => l match {
-        case Lineage(name, _, _, _, _, _, true, _, _) => //SharedParent
-          l.copy(addPort=Some((portNames(name), DecWire)))
-        case Lineage(name, _, _, _, true, _, _, _, _) => //SourceParent
-          l.copy(addPort=Some((portNames(name), DecOutput)))
-        case Lineage(name, _, _, _, _, true, _, _, _) => //SinkParent
-          l.copy(addPort=Some((portNames(name), DecInput)))
-        case Lineage(name, _, _, true, _, _, _, _, _) => //Sink
-          l.copy(addPort=Some((portNames(name), DecInput)))
-        case l => l
-      }),
-      ((l: Lineage) => l match {
-        case Lineage(name, _, true, _, _, _, _, _, _) => //Source
-          val tos = Seq(s"${portNames(name)}")
-          val from = compName
-          l.copy(cons = l.cons ++ tos.map(t => (t, from)))
-        case Lineage(name, _, _, _, true, _, _, _, _) => //SourceParent
-          val tos = Seq(s"${portNames(name)}")
-          val from = l.children.filter { case (i, c) => c.sourceParent }.map { case (i, c) => s"$i.${portNames(c.name)}" }.head
-          l.copy(cons = l.cons ++ tos.map(t => (t, from)))
-        case l => l
-      }),
-      ((l: Lineage) => l match {
-        case Lineage(name, _, _, _, _, true, _, _, _) => //SinkParent
-          val tos = l.children.filter { case (i, c) => (c.sinkParent || c.sink) && !c.sourceParent } map { case (i, c) => s"$i.${portNames(c.name)}" }
-          val from = s"${portNames(name)}"
-          l.copy(cons = l.cons ++ tos.map(t => (t, from)))
-        case l => l
-      })
-    )
-    funs.foldLeft(lin)((l, fun) => fun(l))
+  def namedName(n: Named): String = {
```

Comment:
Since this is extracting the name of the module, I think `moduleName` might be a better name than `namedName`?

Reply:
This makes much more sense. Changed to `getModuleName`.